src/avl_tree.jl

```diff
@@ -0,0 +1,259 @@
+        node = node.leftChild
+    end
+    return node
+end
+
+"""
+    search_node(tree, key)
+
+Returns the last visited node, while traversing through in binary-search-tree fashion looking for `key`.
+"""
+search_node(tree, key)
+
+function search_node(tree::AVLTree{K}, d::K) where K
+    prev = nothing
+    node = tree.root
+    while node != nothing && node.data != nothing && node.data != d
+
+        prev = node
+        if d < node.data 
+            node = node.leftChild
+        else
+            node = node.rightChild
+        end
+    end
+    
+    return (node == nothing) ? prev : node
+end
+
+"""
+    haskey(tree, key)
+
+Returns true if `key` is present in the `tree`, else returns false.    
+"""
+haskey(tree, key)
+
+function Base.haskey(tree::AVLTree{K}, d::K) where K 
+    (tree.root == nothing) && return false
+    node = search_node(tree, d)
+    return (node.data == d)
+end
+
+Base.in(key, tree::AVLTree) = haskey(tree, key)
+
+"""
+    insert!(tree, key)
+
+Inserts `key` in the `tree` if it is not present.
+"""
+insert!(tree, key)
+
+function Base.insert!(tree::AVLTree{K}, d::K) where K
+
+    function insert_node(node::Union{AVLTreeNode, Nothing}, key)
+        if node == nothing
+            return AVLTreeNode{K}(key)
+        elseif key < node.data
+            node.leftChild = insert_node(node.leftChild, key)
+        else
+            node.rightChild = insert_node(node.rightChild, key)
+        end
+        
+        node.height = fix_height(node)
+        balance = get_balance(node)
+        
+        if balance > 1
+            if key < node.leftChild.data
+                return right_rotate(node)
+            else
+                node.leftChild = left_rotate(node.leftChild)
+                return right_rotate(node)
+            end
+        end
+
+        if balance < -1
+            if key > node.rightChild.data
+                return left_rotate(node)
+            else
+                node.rightChild = right_rotate(node.rightChild)
+                return left_rotate(node)
+            end
+        end
+
+        return node
+    end
+
+    haskey(tree, d) && return tree
+
+    tree.root = insert_node(tree.root, d)
+    tree.count += 1
+    return tree
+end
+
+"""
+    push!(tree, key)
+
+Inserts `key` in the `tree` if it is not present.
+"""
+function Base.push!(tree::AVLTree{K}, key0) where K
+    key = convert(K, key0)
+    insert!(tree, key)
+end
+
+function Base.delete!(tree::AVLTree{K}, d::K) where K
+
+    function delete_node(node::Union{AVLTreeNode, Nothing}, key)
+        if node == nothing || node.data == nothing
+            return nothing
+        elseif key < node.data
+            node.leftChild = delete_node(node.leftChild, key)
+        elseif key > node.data
+            node.rightChild = delete_node(node.rightChild, key)
+        else
+            if node.leftChild == nothing
+                temp = node.rightChild
+                node = nothing
+                return temp
```

Comment:
consider rather than calling this `temp`
a more explictit name like `result`, `return_value` or similar.